lib/services/cordova-migration-service.ts

```diff
@@ -134,6 +143,153 @@ export class CordovaMigrationService implements ICordovaMigrationService {
+					if(shouldUpdateFramework) {
+						this.onFrameworkVersionChanging(selectedFramework.Version).wait();
+						this.$project.projectData.FrameworkVersion = selectedFramework.Version;
+					} else {
+						this.$errors.failWithoutHelp("Unable to set Windows Phone %s as the target SDK. Migrate to Apache Cordova 3.7.0 or later and try again.", newVersion);
+					}
+				}
+			}
+		}).future<void>()();
+	}
+
+	public onFrameworkVersionChanging(newVersion: string): IFuture<void> {
+		return ((): void => {
+			if(newVersion === this.$project.projectData.FrameworkVersion) {
+				return;
+			}
+
+			this.$project.ensureAllPlatformAssets().wait();
+
+			if(this.$project.projectData.WPSdk && helpers.versionCompare(this.$project.projectData.WPSdk, "8.0") > 0 && helpers.versionCompare(newVersion, "3.7.0") < 0) {
+				let shouldUpdateWPSdk = this.$prompter.confirm(`You cannot build with the Windows Phone ${this.$project.projectData.WPSdk} SDK with the currently selected target version of Apache Cordova. Do you want to switch to Windows Phone 8.0 SDK?`).wait();
+				if(shouldUpdateWPSdk) {
+					this.onWPSdkVersionChanging("8.0").wait();
+					this.$project.projectData.WPSdk = "8.0";
+				} else {
+					this.$errors.failWithoutHelp("Unable to set %s as the target Apache Cordova version. Set the target Windows Phone SDK to 8.0 and try again.", newVersion);
+				}
+			}
+
+			let versionDisplayName = this.getDisplayNameForVersion(newVersion).wait();
+			this.$logger.info("Migrating to Cordova version %s", versionDisplayName);
+			let oldVersion = this.$project.projectData.FrameworkVersion;
+			
+			let allMarketplacePlugins: string[] = [];
+			_.each(this.$project.configurations, configuration => {
+				let pluginsList = <string[]>this.$project.getProperty("CorePlugins", configuration);
+				let marketplacePlugins = _.filter(pluginsList, plugin => _.contains(plugin, '@'));
+				allMarketplacePlugins = _.union(allMarketplacePlugins, marketplacePlugins);
+			});
+			this.invalidPlugins = _.filter(allMarketplacePlugins, plugin => {
+				let pluginBasicInformation = this.$pluginsService.getPluginBasicInformation(plugin);
+				return !this.$pluginsService.isPluginSupported(pluginBasicInformation.name, pluginBasicInformation.version, newVersion);
+			});
+
+			if (this.invalidPlugins.length > 0) {
+				this.promptUserForInvalidPluginsAction(this.invalidPlugins, newVersion).wait();
+			}
+
+			_.each(this.$project.configurations, (configuration: string) => {
+				let oldPluginsList = this.$project.getProperty("CorePlugins", configuration);
+				let newPluginsList = this.migratePlugins(oldPluginsList, oldVersion, newVersion).wait();
+				this.$logger.trace("Migrated core plugins to: ", helpers.formatListOfNames(newPluginsList, "and"));
+				this.$project.setProperty("CorePlugins", newPluginsList, configuration);
+			});
+
+			let backedUpFiles: string[] = [],
```

Comment:
Consider moving the code from this line bellow including the try .. finally into a dedicated function.